### ISSUE TITLE
option not to remove existing dropin files.

### ISSUE THREAD

**Issue by hboetes** [voxpupuli/puppet-systemd#504]
After creating this dropin file, noticed puppet removed the existing dropin created by freeipa

```
    systemd::manage_dropin { 'puppet_override.conf':
      ensure        => present,
      unit          => 'pki-tomcatd@pki-tomcat.service',
      service_entry  => {
        'Restart'    => 'on-failure',
        'RestartSec' => '60',
      },
    }
```

Please add the option not to remove existing dropin files.

**Comment by kenyon** [comment]
I don't see anywhere in this module's code that would cause removal of dropin files that aren't puppet managed.

**Comment by hboetes** [comment]
That's what I saw in the puppet output, it removed the other dropin straight away.

**Comment by traylenator** [comment]
That's correct , as soon as one drop in is managed the directory is purged via:

https://github.com/voxpupuli/puppet-systemd/blob/master/manifests/dropin_file.pp#L56

Looks to be a setting `systemd::purge_dropin_dirs` 

https://github.com/voxpupuli/puppet-systemd/blob/master/manifests/init.pp#L287

Never used it but presumably works. Looks to be global setting only.

I think for the odd case where this has been a problem I normaly just add a trivial file  with `replace => false` on the file I don't
want to be deleted.

Steve.

**Comment by TheMeier** [comment]
Hm, not a big fan. It smell like multiple different tools configure one resource (the service). 
My gut feeling is: If you manage the drop-ins for a service in puppet, you should also manage the service itself (and thus all it's dropins)

Having 2 switches (global and per-instance) needs some additional instance. And how would you decide this:

```
systemd::manage_dropin { 'myconf1.conf':
  ensure        => present,
  unit             => 'myservice.service',
  purge_dir   => true,
}

systemd::manage_dropin { 'myconf2.conf':
  ensure        => present,
  unit             => 'myservice.service',
  purge_dir   => false,
}
```

**Comment by traylenator** [comment]
> Having 2 switches (global and per-instance) needs some additional logic. And how would you decide this:
> 
> ```
> systemd::manage_dropin { 'myconf1.conf':
>   ensure        => present,
>   unit             => 'myservice.service',
>   purge_dir   => true,
> }
> 
> systemd::manage_dropin { 'myconf2.conf':
>   ensure        => present,
>   unit             => 'myservice.service',
>   purge_dir   => false,
> }
> ```

It would come up as a conflict between the two directory instances that are already created within ensure_resource.

That's an okay result I'd say.

**Comment by hboetes** [comment]
In my case, freeipa is installing a dropin as well, and I don't want to override that, since it may change with an upcoming release. For your proposed problem situation, anything goes, but a check for said situation would be nice indeed, ideally it would result in a conflict indeed.

The per instance setting should override the global setting, I'd say.

**Comment by TheMeier** [comment]
I still would argue that your puppet managed dropins might contradict the package provided ones, potentially without you noticing on a package upgrade. So if you **really** want to have full control you should go with either one any not mix them. Implicit definitions a potential recipe for disaster.  

Disclaimer: just a personal opinion at this moment in time :wink:

**Comment by traylenator** [comment]
> In my case, freeipa is installing a dropin as well, and I don't want to override that, since it may change with an upcoming release. For your proposed problem situation, anything goes, but a check for said situation would be nice indeed, ideally it would result in a conflict indeed.
> 
> The per instance setting should override the global setting, I'd say.

Just add:

```puppet
file{'/etc/systemd/system/freeipa.service.d/file_from_rpm.conf':
     ensure => file,
     replace=> false,
     require => Package['freeipa'],
}
```
To protect the file.

**Comment by hboetes** [comment]
Thanks for your suggestion. Feels a bit hackish, though.

**Comment by TheMeier** [comment]
@hboetes is the provided solution good enough to consider it done though?

**Comment by hboetes** [comment]
Since there is no ideal solution, yes.
